misk/src/main/kotlin/misk/web/interceptors/ConcurrencyLimitsInterceptor.kt

```diff
@@ -0,0 +1,79 @@
+package misk.web.interceptors
+
+import com.netflix.concurrency.limits.Limiter
+import com.netflix.concurrency.limits.limiter.SimpleLimiter
+import misk.Action
+import misk.exceptions.StatusCode
+import misk.logging.getLogger
+import misk.web.AvailableWhenDegraded
+import misk.web.ConcurrencyLimitsOptIn
+import misk.web.NetworkChain
+import misk.web.NetworkInterceptor
+import java.time.Clock
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.full.findAnnotation
+
+/**
+ * Detects degraded behavior and sheds requests accordingly. Internally this uses adaptive limiting
+ * as implemented by Netflix's [concurrency-limits][concurrency_limits] library. See
+ * [AvailableWhenDegraded] for further documentation.
+ *
+ * [concurrency_limits]: https://github.com/Netflix/concurrency-limits/
+ */
+internal class ConcurrencyLimitsInterceptor internal constructor(
+  private val actionName: String,
+  private val limiter: Limiter<String>
+) : NetworkInterceptor {
+
+  override fun intercept(chain: NetworkChain) {
+    val listener: Limiter.Listener? = limiter.acquire(actionName).orElse(null)
+
+    if (listener == null) {
+      logger.error { "concurrency limits interceptor shedding $actionName" }
+      chain.httpCall.statusCode = StatusCode.SERVICE_UNAVAILABLE.code
+      chain.httpCall.takeResponseBody()?.use { sink ->
+        sink.writeUtf8("service unavailable")
+      }
+      return
+    }
+
+    try {
+      chain.proceed(chain.httpCall)
+    } catch (unexpected: Throwable) {
+      listener.onIgnore()
+      throw unexpected
+    }
+
+    try {
+      when (chain.httpCall.statusCode) {
+        in 400 until 500 -> listener.onIgnore()
+        in 500 until 600 -> listener.onDropped() // Count 5XX errors as drops.
+        else -> listener.onSuccess()
+      }
+    } catch (e: IllegalArgumentException) {
+      // Service container does a ignores an RTT == 0 exception here. Is this still necessary?
+      logger.info { "ignoring concurrency-limits exception: $e" }
```

Comment:
warn?

Reply:
I actually changed it the opposite way to `debug` because it was very chatty.

Reply:
(And I’m gonna report a bug upstream.)